database/db-driver/greenplum/db-assign_sequential_id

```diff
@@ -1,97 +1,93 @@
 #!/usr/bin/env bash
 # db-assign_sequential_id -- Assigns a unique integer to every row for a table using PostgreSQL sequence generator
 # > eval "$(db-parse "$url")"
-# > db-assign_sequential_id TABLE COLUMN BEGIN_ID
+# > db-assign_sequential_id TABLE COLUMN BEGIN_ID [INCREMENT]
 ##
 set -euo pipefail
 
 [[ $# -gt 0 ]] || usage "$0" "Missing TABLE"
 [[ $# -gt 1 ]] || usage "$0" "Missing COLUMN"
 [[ $# -gt 2 ]] || usage "$0" "Missing BEGIN_ID"
-Table=$1 Column=$2 BeginId=$3
+Table=$1 Column=$2 BeginId=$3 Increment=${4:-1}
 
 # Use Greenplum PL/pgSQL and PL/Python UDFs to assign IDs fast
 # See: http://www.postgresql.org/docs/8.2/static/sql-createlanguage.html
 # See: http://www.postgresql.org/docs/8.2/static/plpgsql-overview.html
 # See: http://www.postgresql.org/docs/8.2/static/plpython-funcs.html
-db-execute "CREATE LANGUAGE plpgsql;" || true
-db-execute "CREATE LANGUAGE plpythonu;" || true
-db-execute "
-    CREATE OR REPLACE FUNCTION clear_count_1(sid int) RETURNS int AS
-    \$\$
-    if '__count_1' in SD:
-      SD['__count_1'] = -1
-      return 1
-    return 0
+if db-supports_pg_lang "plpgsql" && db-supports_pg_lang "plpythonu"; then
+    db-execute "
+    CREATE OR REPLACE FUNCTION fast_seqassign_init()
+    RETURNS INT AS \$\$
+      if 'cumulative_cnt' in SD:
+        SD.pop('cumulative_cnt')
+      return 0
     \$\$ LANGUAGE plpythonu;
 
-    CREATE OR REPLACE FUNCTION updateid(startid bigint, sid int, sids int[], base_ids bigint[], base_ids_noagg bigint[]) RETURNS bigint AS
-    \$\$
-    if '__count_1' in SD:
-      a = SD['__count_2']
-      b = SD['__count_1']
-      SD['__count_2'] = SD['__count_2'] - 1
-      if SD['__count_2'] < 0:
-        SD.pop('__count_1')
-      return startid+b-a
-    else:
-      for i in range(0, len(sids)):
-        if sids[i] == sid:
-          SD['__count_1'] = base_ids[i] - 1
-          SD['__count_2'] = base_ids_noagg[i] - 1
-      a = SD['__count_2']
-      b = SD['__count_1']
-      SD['__count_2'] = SD['__count_2'] - 1
-      if SD['__count_2'] < 0:
-        SD.pop('__count_1')
-      return startid+b-a
-
+    CREATE OR REPLACE FUNCTION fast_seqassign_next(startid BIGINT, increment BIGINT, this_gpsid INT, gpsids INT[], cumulative_cnts BIGINT[], cnts BIGINT[])
+    RETURNS BIGINT AS \$\$
+      if 'cumulative_cnt' not in SD:
+        for gpsid, cumulative_cnt, cnt in zip(gpsids, cumulative_cnts, cnts):
+          if gpsid == this_gpsid:
+            # assignment is done using the cumulative count for this segment and a decrementing counter from the size of this segment down to one
+            SD['cumulative_cnt'] = cumulative_cnt
+            SD['remaining'] = cnt
+            break
+      if 'cumulative_cnt' in SD and SD['remaining'] > 0:
+        id = startid + increment * (SD['cumulative_cnt'] - SD['remaining'])
+        SD['remaining'] = SD['remaining'] - 1
+        if SD['remaining'] <= 0:
+          SD.pop('cumulative_cnt')
```

Comment:
`SD['remaining'] == 0` is the acceptable case. After that, there should be no more calls. However, if you do `SD.pop` here, the next call would enter the above `if 'cumulative_cnt' not in SD:` branch instead of the `raise` statement.

If you just remove these two lines, the next call would go to the `raise` statement.